src/drivers/pid_driver/driver.cpp

```diff
@@ -0,0 +1,325 @@
+#include "driver.h"
+
+#include <cmath>
+#include <iostream>
+#ifndef M_PI
+#define M_PI (3.14159265358979323846)
+#endif
+
+const float Driver::MAX_UNSTUCK_ANGLE = 30.0 / 180.0 * PI; /* [radians] */
+const float Driver::UNSTUCK_TIME_LIMIT = 2.0;              /* [s] */
+const float Driver::MAX_UNSTUCK_SPEED = 5.0;               /* [m/s] */
+const float Driver::MIN_UNSTUCK_DIST = 3.0;                /* [m] */
+const float Driver::G = 9.81;                              /* [m/(s*s)] */
+const float Driver::FULL_ACCEL_MARGIN = 1.0;               /* [m/s] */
+const float Driver::SHIFT = 0.9;                           /* [-] (% of rpmredline) */
+const float Driver::SHIFT_MARGIN = 4.0;                    /* [m/s] */
+const float Driver::ABS_SLIP = 0.9;                        /* [-] range [0.95..0.3] */
+const float Driver::ABS_MINSPEED = 3.0;                    /* [m/s] */
+const float Y_DIST_TO_MIDDLE = 5.0;
+const float GOAL_POS_Y = -20;
+const float GOAL_POS_X = 0;
+const float MIN_DIST = 4;
+
+Driver::Driver(int index) {
+    float dt = 0.02;
+    float Kp = -0.3;
+    float Kd = -0.2;
+    float Ki = -0.001;
+    _pidAcc = PidAcc(dt, Kp, Kd, Ki);
+    Kp = -0.1;
+    Kd = -1.0;
+    Ki = -0.005;
+    float G1 = 0.2;
+    float G2 = 0.8;
+    _pidSteer = PidSteer(dt, Kp, Kd, Ki, G1, G2, 12);
+    INDEX = index;
+}
+
+/* Called for every track change or new race. */
+void Driver::initTrack(tTrack *t, void *carHandle, void **carParmHandle, tSituation *s) {
+    track = t;
+    *carParmHandle = NULL;
+}
+
+/* Start a new race. */
+void Driver::newRace(tCarElt *car, tSituation *s) {
+    this->car = car;
+    CARMASS = GfParmGetNum(car->_carHandle, SECT_CAR, PRM_MASS, NULL, 1000.0);
+    initCa();
+    initCw();
+
+    MAX_UNSTUCK_COUNT = int(UNSTUCK_TIME_LIMIT / RCM_MAX_DT_ROBOTS);
+    stuck = 0;
+
+    /* initialize the list of opponents */
+    opponents = new Opponents(s, this);
+    opponent = opponents->getOpponentPtr();
+}
+
+/* Drive during race. */
+void Driver::drive(tCarElt *car, tSituation *s) {
+    update(car, s);
+    // std::cout << "Other cars ----" << std::endl;
+    for (int i = 0; i < opponents->getNOpponents(); i++) {
+        // std::cout << "SpeedY " << i << ": " << getSpeed() << std::endl;
+        // std::cout << "SpeedDiffX " << i << ": " << getOpponentSpeedDiffX(opponent[i]) << std::endl;
+        // std::cout << "SpeedDiffY " << i << ": " << getOpponentSpeedDiffY(opponent[i]) << std::endl;
+        // std::cout << "DistanceY " << i << ": " << getOpponentDistanceY(opponent[i]) << std::endl;
+        // std::cout << "DistanceX " << i << ": " << getOpponentDistanceX(opponent[i]) << std::endl;
+        std::cout << getOpponentDistanceX(opponent[i]) << ", " << getOpponentDistanceY(opponent[i]) << std::endl;
```

Comment:
please remove this print statement

Reply:
I left the print statements and just commented them out. They might be usefull in the future again, in case someone wants to work on the project again or we need to fix more issues.